### PR TITLE
WebSocket:  getAllKeys, runCode. Frontend: UI to show list of state and upload file.

### PR DESCRIPTION
**Pull request by ChristopherTrimboli**

### CODE REVIEW

File: examples/desktop.html [hunk 1]
```diff
@@ -198,17 +233,52 @@ <h1>Connect to room</h1>
...
+      name: file.name,
+      type: file.type,
+      size: file.size
+    };
+
+    const reader = new FileReader();
+    reader.addEventListener("load", () => {
+      fileObject.src = reader.result;
+      console.log(fileObject);
+      channelConnection.runScript(fileObject.name, fileObject);
```
Comment (avaer) [inline examples/desktop.html]: Shouldn't this be a promise?

Comment (ChristopherTrimboli) [reply examples/desktop.html]: no i don't think so because you receive the reply in a listener not as a return response like that using websocket.notify()

Comment (ChristopherTrimboli) [reply examples/desktop.html]: https://github.com/webaverse/metartc2/pull/1/files#diff-82fa352fb7ce5eebb62f4eda8530920cR126-R130

Comment (avaer) [reply examples/desktop.html]: You'd want to know when the script is done running though. That's probably what you care about as the caller, not any events.

Comment (avaer) [reply examples/desktop.html]: The only events we'd want to listen for are the ones that say a chunk got updated. But that is in addition to knowing when our script run completed.

File: examples/desktop.html [hunk 2]
```diff
@@ -1,31 +1,66 @@
...
+    </div>
   </form>
-<script type=module>
+  <script type=module>
 import THREE from '../three.module.js';
-import {XRChannelConnection} from '../xrrtc.js';
+import { XRChannelConnection } from '../xrrtc.js';
 
-const presenceHost = 'wss://rtc.exokit.org:4443';
+const presenceHost = 'wss://127.0.0.1:4443';
```
Comment (avaer) [inline examples/desktop.html:63]: This will break when deployed.

Comment (ChristopherTrimboli) [reply examples/desktop.html:63]: good catch

File: examples/desktop.html [hunk 3]
```diff
@@ -198,17 +233,64 @@ <h1>Connect to room</h1>
...
+    e.data.forEach(key => {
+      const line = document.createElement("li");
+      const text = document.createTextNode(key);
+      line.appendChild(text);
+      list.appendChild(line);
+    }) 
+
+  }, {once: true});
+
+  setTimeout(() => {
```
Comment (avaer) [inline examples/desktop.html]: Why this timeout?

File: examples/desktop.html (hunk 1 above)
Comment (ChristopherTrimboli) [reply examples/desktop.html]: outdated

File: examples/desktop.html (hunk 3 above)
Comment (ChristopherTrimboli) [reply examples/desktop.html]: 7030696

File: examples/desktop.html (hunk 2 above)
Comment (ChristopherTrimboli) [reply examples/desktop.html:63]: a66ba4a

**Comment by ChristopherTrimboli** [comment]
ready for review, includes demo code using new API format

File: client/RoomClient.js [hunk 4]
```diff
@@ -843,13 +843,70 @@ export default class RoomClient extends EventTarget
...
+					case 'runCode':
+					{
+						console.log('Room.js, runCode:', notification)
 						this.dispatchEvent(new MessageEvent(notification.method, {
 							data: notification.data,
 						}));
+						break;
+					}
 
+					case 'uploadBinary':
```
Comment (avaer) [inline client/RoomClient.js]: If this isn't used we should get rid of it.

File: client/RoomClient.js [hunk 5]
```diff
@@ -843,13 +843,70 @@ export default class RoomClient extends EventTarget
...
+					case 'edit':
+					{
+						console.log('Room.js, edit:', notification)
+						this.dispatchEvent(new MessageEvent(notification.method, {
+							data: notification.data,
+						}));
+						break;
+					}
+
+					case 'getAllKeys':
```
Comment (avaer) [inline client/RoomClient.js:868]: What's the difference between keys and files? Can that be written into the API names?

File: client/RoomClient.js [hunk 6]
```diff
@@ -843,13 +843,70 @@ export default class RoomClient extends EventTarget
...
+					case 'getFile':
+					{
+						console.log('Room.js, getFile:', notification)
+						this.dispatchEvent(new MessageEvent(notification.method, {
+							data: notification.data,
+						}));
+						break;
+					}
+
+					case 'edit':
```
Comment (avaer) [inline client/RoomClient.js]: What is an edit vs runCode? Can we delete this?

File: client/RoomClient.js [hunk 7]
```diff
@@ -1617,16 +1674,46 @@ export default class RoomClient extends EventTarget
         	key,
         	value,
         });
-    }
+	}
+	
+	getFile(key) {
```
Comment (avaer) [inline client/RoomClient.js]: Shouldn't this be a promise that resolves when the file is downloaded? Also... could this be a separate rest call to not clog the websocket?

File: client/RoomClient.js (hunk 5 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js:868]: 👍

File: client/RoomClient.js (hunk 4 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js]: removed unused apis

File: client/RoomClient.js (hunk 5 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js:868]: we should just talk in "keys"

File: client/RoomClient.js (hunk 7 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js]: outdated

File: client/RoomClient.js (hunk 6 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js]: outdated, i removed edit

File: client/RoomClient.js (hunk 5 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js:868]: no more files just keys

File: client/RoomClient.js [hunk 8]
```diff
@@ -1617,16 +1647,21 @@ export default class RoomClient extends EventTarget
...
+	
+	getAllKeys() {
+	    if (this._closed)
+	        return;
 
-    deleteState(key, value) {
+        this._protoo.notify('getAllKeys');
+	}
+	
+	runCode(obj) {
```
Comment (avaer) [inline client/RoomClient.js:1659]: How does the client know the run completed? Should this be a promise?

File: xrrtc.js [hunk 9]
```diff
@@ -141,8 +146,12 @@ class XRChannelConnection extends EventTarget {
     this.dialogClient.setState(key, value);
   }
 
-  deleteState(key) {
-    this.dialogClient.deleteState(key);
+  getAllKeys() {
+    this.dialogClient.getAllKeys();
+  }
+
+  runCode(obj) {
```
Comment (avaer) [inline xrrtc.js]: Another place that might need to be a Promise.

File: client/RoomClient.js (hunk 8 above)
Comment (ChristopherTrimboli) [reply client/RoomClient.js:1659]: did it in a event listener

File: xrrtc.js (hunk 9 above)
Comment (ChristopherTrimboli) [reply xrrtc.js]: hmm what about now?

Comment (avaer) [reply xrrtc.js]: This is still not async?

Comment (ChristopherTrimboli) [reply xrrtc.js]: 1f02ce4

**Review by avaer — APPROVED** [review]

**Pull request merged**